File: leetcode/valid-parentheses.js
/**
 * @param {string} s
 * @return {boolean}
 */
const pairs = ["()", "{}", "[]"];
const leftMap = pairs.reduceRight((cur, pair) => cur.set(pair[0], pair[1]), new Map())
const rightMap = pairs.reduceRight((cur, pair) => cur.set(pair[1], pair[0]), new Map())
var isValid = function (s) {
  if (s.length % 2 == 1) return false;

  let leftStack = [];
  for (let i = 0; i < s.length; i++) {
    if (leftStack.length > s.length - i) return false;

    let chr = s[i];
    if (leftMap.has(chr)) leftStack.push(chr);
    else if (rightMap.has(chr)) {
      let leftChr = leftStack.pop();
      if (leftMap.get(leftChr) !== chr) return false;
    }
  }
  return leftStack.length == 0;
};

File: leetcode/valid-parentheses.cpp
#include "string"
#include "vector"
#include "map"
#include "stack"
using namespace std;

const string pairs[] = {"{}", "[]", "()"};
map<char, char> leftMap, rightMap;
void initMap()
{
  for (string pair : pairs)
  {
    leftMap.emplace(pair[0], pair[1]);
    rightMap.emplace(pair[1], pair[0]);
  }
}

class Solution
{
public:
  bool isValid(string s)
  {
    if (s.length() % 2 == 1)
      return false;
    stack<char> leftStack;
    initMap();

    for (int i = 0; i < s.length(); i++)
    {
      char chr = s[i];
      if (leftStack.size() > s.length() - i)
        return false;
        
      if (leftMap.find(chr) != leftMap.end())
        leftStack.push(chr);
      else if (rightMap.find(chr) != rightMap.end())
      {
        if (leftStack.empty())
          return false;
        char leftChr = leftStack.top();
        leftStack.pop();
        if (leftMap.at(leftChr) != chr)
          return false;
      }
    }
    return leftStack.empty();
  }
};